inference/core/workflows/core_steps/models/foundation/openai.py

```diff
@@ -0,0 +1,407 @@
+import asyncio
+import base64
+import json
+import re
+from typing import Any, Dict, List, Literal, Optional, Type, Union
+
+from openai import AsyncOpenAI
+from pydantic import BaseModel, ConfigDict, Field
+
+from inference.core.env import WORKFLOWS_REMOTE_EXECUTION_MAX_STEP_CONCURRENT_REQUESTS
+from inference.core.managers.base import ModelManager
+from inference.core.utils.image_utils import encode_image_to_jpeg_bytes, load_image
+from inference.core.workflows.constants import PARENT_ID_KEY, ROOT_PARENT_ID_KEY
+from inference.core.workflows.core_steps.common.entities import StepExecutionMode
+from inference.core.workflows.entities.base import (
+    Batch,
+    OutputDefinition,
+    WorkflowImageData,
+)
+from inference.core.workflows.entities.types import (
+    BATCH_OF_DICTIONARY_KIND,
+    BATCH_OF_IMAGE_METADATA_KIND,
+    BATCH_OF_PARENT_ID_KIND,
+    BATCH_OF_STRING_KIND,
+    DICTIONARY_KIND,
+    STRING_KIND,
+    WILDCARD_KIND,
+    ImageInputField,
+    StepOutputImageSelector,
+    WorkflowImageSelector,
+    WorkflowParameterSelector,
+)
+from inference.core.workflows.prototypes.block import (
+    BlockResult,
+    WorkflowBlock,
+    WorkflowBlockManifest,
+)
+from inference_sdk.http.utils.iterables import make_batches
+
+NOT_DETECTED_VALUE = "not_detected"
+JSON_MARKDOWN_BLOCK_PATTERN = re.compile(r"```json\n([\s\S]*?)\n```")
+
+
+class LMMConfig(BaseModel):
+    max_tokens: int = Field(default=450)
+    gpt_image_detail: Literal["low", "high", "auto"] = Field(default="auto")
+    gpt_model_version: str = Field(default="gpt-4o")
+
+
+LONG_DESCRIPTION = """
+Ask a question to OpenAI's GPT-4 with Vision model.
+
+You can specify arbitrary text prompts to the OpenAIBlock.
+
+You need to provide your OpenAI API key to use the GPT-4 with Vision model. 
+
+_This model was previously part of the LMM block._
+"""
+
+
+class BlockManifest(WorkflowBlockManifest):
+    model_config = ConfigDict(
+        json_schema_extra={
+            "name": "OpenAI",
+            "short_description": "Run OpenAI's GPT-4 with Vision",
+            "long_description": LONG_DESCRIPTION,
+            "license": "Apache-2.0",
+            "block_type": "model",
+            "search_keywords": ["LMM", "ChatGPT"],
+        }
+    )
+    type: Literal["OpenAI"]
+    images: Union[WorkflowImageSelector, StepOutputImageSelector] = ImageInputField
+    prompt: Union[WorkflowParameterSelector(kind=[STRING_KIND]), str] = Field(
+        description="Text prompt to the OpenAI model",
+        examples=["my prompt", "$inputs.prompt"],
+    )
+    api_key: Union[WorkflowParameterSelector(kind=[STRING_KIND]), Optional[str]] = (
```

Comment:
This is only thought - we use `api_key` to denote our `api_key`, I think it would be good idea to call this `open_ai_api_key` to avoid name clash, and also it would fit well with other parameter below `open_ai_model`